File: algocoding/leetcode/add_two_numbers.cpp
/*******************************************************************************
You are given two linked lists representing two non-negative numbers. The digits 
are stored in reverse order and each of their nodes contain a single digit. Add 
the two numbers and return it as a linked list.

Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
Output: 7 -> 0 -> 8

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 *//*
class Solution {
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        
    }
};

*******************************************************************************/





#include <cstddef>

struct ListNode 
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) 
{
    ListNode lhead(-1);
    int rem = 0,a=0,b=0;
    ListNode *ltail = &lhead;	
    ListNode *la = l1;
    ListNode *lb = l2;

	while(NULL != la || NULL != lb)
	{
	    a = (NULL==la ? 0:la->val);
	    b = (NULL==lb ? 0:lb->val);
		ltail->next = new ListNode((a + b + rem) % 10);
		rem = (a + b + rem) / 10;
		ltail = ltail->next;
		la = (NULL==la ? NULL:la->next);
		lb = (NULL==lb ? NULL:lb->next);
	}
		
    if (rem > 0)
        ltail->next = new ListNode(rem);
	
    return lhead.next;
}

};

int main()
{
    return 0;
}

File: algocoding/leetcode/add_two_numbers.py
#!/usr/bin/env python
#coding=utf-8

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @return a ListNode
    def addTwoNumbers(self, l1, l2):
        # method 1: 14 Lines
        dummy, flag = ListNode(0), 0
        head = dummy
        while flag or l1 or l2:
            node = ListNode(flag)
            if l1:
                node.val += l1.val
                l1 = l1.next
            if l2:
                node.val += l2.val
                l2 = l2.next
            flag = node.val / 10
            node.val %= 10
            head.next = node  
            head = head.next  # head.next, head = node, node
        return dummy.next
        
#Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
#Output: 7 -> 0 -> 8
a1 = ListNode(2)
a2 = ListNode(4)
a3 = ListNode(3)
b1 = ListNode(5)
b2 = ListNode(6)
b3 = ListNode(4)
a1.next = a2
a2.next = a3
b1.next = b2
b2.next = b3

solu = Solution()
head = ListNode(-1)
head.next = solu.addTwoNumbers(a1,b1)

def show_list(head):
    while head:
	    print head.val
	    head = head.next
		
show_list(head.next)

'''
asdfasdf
asdfasdfdsfasdf
'''
